ludwig/data/preprocessing.py

```diff
@@ -940,79 +941,92 @@ def build_dataset(
 
 def build_metadata(dataset_df, features, global_preprocessing_parameters):
     metadata = {}
+
     for feature in features:
-        if PREPROCESSING in feature:
-            preprocessing_parameters = merge_dict(
-                global_preprocessing_parameters[feature[TYPE]],
-                feature[PREPROCESSING]
-            )
-        else:
-            preprocessing_parameters = global_preprocessing_parameters[
-                feature[TYPE]
-            ]
 
-        # deal with encoders that have fixed preprocessing
-        if 'encoder' in feature:
-            encoders_registry = get_from_registry(
-                feature[TYPE],
-                input_type_registry
-            ).encoder_registry
-            encoder_class = encoders_registry[feature['encoder']]
-            if hasattr(encoder_class, 'fixed_preprocessing_parameters'):
-                encoder_fpp = encoder_class.fixed_preprocessing_parameters
+        if HASH not in feature:
+            feature[HASH] = compute_feature_hash(feature)
+
+        if feature[HASH] not in metadata:
 
+            if PREPROCESSING in feature:
                 preprocessing_parameters = merge_dict(
-                    preprocessing_parameters,
-                    resolve_pointers(encoder_fpp, feature, 'feature.')
+                    global_preprocessing_parameters[feature[TYPE]],
+                    feature[PREPROCESSING]
                 )
+            else:
+                preprocessing_parameters = global_preprocessing_parameters[
+                    feature[TYPE]
+                ]
+
+            # deal with encoders that have fixed preprocessing
+            if 'encoder' in feature:
+                encoders_registry = get_from_registry(
+                    feature[TYPE],
+                    input_type_registry
+                ).encoder_registry
+                encoder_class = encoders_registry[feature['encoder']]
+                if hasattr(encoder_class, 'fixed_preprocessing_parameters'):
+                    encoder_fpp = encoder_class.fixed_preprocessing_parameters
+
+                    preprocessing_parameters = merge_dict(
+                        preprocessing_parameters,
+                        resolve_pointers(encoder_fpp, feature, 'feature.')
+                    )
+
+            get_feature_meta = get_from_registry(
+                feature[TYPE],
+                base_type_registry
+            ).get_feature_meta
 
-        get_feature_meta = get_from_registry(
-            feature[TYPE],
-            base_type_registry
-        ).get_feature_meta
-
-        metadata[feature[NAME]] = get_feature_meta(
-            dataset_df[feature[NAME]].astype(str),
-            preprocessing_parameters
-        )
+            metadata[feature[HASH]] = get_feature_meta(
```

Comment:
I see how this gets confusing.  One solution, at the cost of adding more keystrokes, would be to rename as follows:

```
NAME -> INPUT_COLUMN
HASH -> PROCESSED_COLUMN
```

I think that makes the relationship more clear.